src/README.md

```diff
@@ -0,0 +1,180 @@
+	private $container;
+
+	public function __construct( Container $container ) {
+		$this->container = $container;
+	}
+
+	public function my_method() {
+		$state = $this->container->get( Completed_State::class );
+	}
+}
+```
+
+### Using the container within `includes`
+
+`wcpay_get_container(): WCPay\Container` allows files in `includes` to load dependencies from within `src`. The function should never be used within `src`.
+
+Use the `wcpay_get_container()->get( $id )` to load dependencies. This can be done:
+
+- Within constructors whenever possible.
+- Within a separate private method, which has a proper type hint. Otherwise the result of `wcpay_get_container()->get()` is `mixed`, which will prevent IDEs from understanding it correctly.
+
+```php
+use WCPay\Internal\Service\PaymentProcessingService;
+
+class MyClass {
+	/**
+	 * @var PaymentProcessingService
+	 */
+	private $payment_processing_service;
+
+	public function __construct() {
+		$container = wcpay_get_container();
+		$this->payment_processing_service = $container->get( PaymentProcessingService::class );
+	}
+
+	// or...
+
+	/**
+	 * Returns the dependency with a proper type hint.
+	 * @return PaymentProcessingService
+	 */
+	private function get_payment_processing_service() {
+		$container = wcpay_get_container();
+		return $container->get( PaymentProcessingService::class );
+	}
+}
+```
+
+### Registering classes
+
+We use [service providers](Internal/DependencyManagement/AbstractServiceProvider.php) to register our classes with the container. This is done manually in order to avoid having to use [reflection](https://www.php.net/manual/en/book.reflection.php), which may have a performance hit on merchant sites.
+
+All service providers should reside within the `WCPay\Internal\DependencyManagement\ServiceProvider` namespace. Depending on the class you are registering, you can either use an existing service, or create another one for the topic of the class.
+
+Here is the anatomy of a service:
+
+```php
+namespace WCPay\Internal\DependencyManagement\ServiceProvider;
+
+use WCPay\Internal\DependencyManagement\AbstractServiceProvider;
+// None of those exists, they are just examples.
+use WC_Payments_API_Client;
+use WCPay\Internal\Service\SimpleService;
+use WCPay\Internal\Service\ServiceWithDependencies;
+use WCPay\Internal\Request;
+
+class PaymentsServiceProvider extends AbstractServiceProvider {
+	/**
+	 * Contains all provided classes/aliases.
+	 * @var string[]
+	 */
+	protected $provides = [
+		SimpleService::class,
+		SimpleServiceWithDependencies::class,
+		Request::class,
+	];
+
+	/**
+	 * Registers all provided classes.
+	 */
+	public function register(): void {
+		$container = $this->getContainer();
+
+		$container->addShared( SimpleService::class );
+
+		$container->addShared( SimpleServiceWithDependencies::class )
+			->addArgument( ServiceWithDependencies::class);
+
+		$container->add( Request::class )
+			->addArgument( WC_Payments_API_Client::class );
```

Comment:
The readme will be updated in https://github.com/Automattic/woocommerce-payments/pull/6876, which introduces a `LegacyContainer` that has hard-coded ways to access legacy classes. If the approach there is approved, this example will be valid.

Is there any chance that you could take a quick glance at https://github.com/Automattic/woocommerce-payments/pull/6876/files#diff-97c044d71e0d3b85870e78635d1f2de33e6ba482799e5d61ac21485432ab6f93, and comment on the approach before the PR is polished? Note that the idea is similar to Woo's `LegacyContainer` ([ref](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Proxies/LegacyProxy.php#L69)), but instead of accessing legacy instances through `LegacyContainer`, they would be available as direct dependencies. Once moved to `src`, the classes that depend on them will not require modifications.